src/pandaboxlib/pandaboxlib.py

```diff
@@ -6,80 +6,282 @@
 
 import socket
 import sys
-from .save_config import GetLines, read_response, save_state, save_table
-from .save_config import save_metatable, save_metadata
+
+
+class _Design:
+    """Helper class for saving/loading a PandABox block-design
+
+    Based on example client in `PandABlocks-server <https://github.com/PandABlocks/PandABlocks-server/blob/master/python/save-state>`_
+
+    """
```

Comment:
I suggest to add a test which covers this class's methods.

Reply:
Yeah I agree we should probably have some tests for this interface.

Fundamentally, the `_Design` class was just a lazy migration of the code from the `save_config.py` module to the main module, with the addition of a few methods. I'm not super happy with the encapsulation to be very honest; a lot of code — particularly related to comms — is repeated in both `_Design` and `PandA`.

I would therefore propose the following;

* Add tests for the current public facing interfaces of the `PandA` class (to ensure nothing gets broken)
* Add tests for the public facing interfaces of the `_Design` class (to define desired behaviour)
* Refactor both `_Design` and `PandA` classes to be DRYer

What do you think? I made a start on the refactoring already, so I reckon I could probably push some changes within a few days.

Reply:
Yes I think this will go a ways to making the code tidier.

Maybe a nice socket wrapper that both can use... and we have several socket "clients" to devices like dg645 and coboldpc that are currently pretty sloppy and could use some better error handling.

Reply:
Hej! Brief update; the 'quick changes' are of course turning into a full re-write of the library…

Current status;

* [x] Add tests for the current public facing interfaces of the `PandA` class (to ensure nothing gets broken)
* [ ] Add tests for the public facing interfaces of the `_Design` class (to define desired behaviour) — **in progress**
* [ ] Refactor both `_Design` and PandA classes to be DRYer — **in progress**

I haven't yet worked out a way to write a good reusable socket wrapper. Will see if I can come up with something though.
On the other hand, I've now got a nice generic mock socket which might come in handy for testing socket based comms.

Will push a few commits in the coming days so you guys can take a look.